src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -0,0 +1,400 @@
+    inputDimension = std::move(layer.inputDimension);
+    size = std::move(layer.size);
+    higherDimension = std::move(layer.higherDimension);
+    runningMean = std::move(layer.runningMean);
+    runningVariance = std::move(layer.runningVariance);
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+void BatchNormType<MatType>::SetWeights(
+    typename MatType::elem_type* weightsPtr)
+{
+  MakeAlias(weights, weightsPtr, WeightSize(), 1);
+  // Gamma acts as the scaling parameters for the normalized output.
+  MakeAlias(gamma, weightsPtr, size, 1);
+  // Beta acts as the shifting parameters for the normalized output.
+  MakeAlias(beta, weightsPtr + gamma.n_elem, size, 1);
+}
+
+template<typename MatType>
+void BatchNormType<MatType>::CustomInitialize(
+    MatType& W,
+    const size_t elements)
+{
+  if (elements != 2 * size) {
+    throw std::invalid_argument("BatchNormType::CustomInitialize(): wrong "
+        "elements size!"); 
+  }
+  MatType gammaTemp;
+  MatType betaTemp;
+  // Gamma acts as the scaling parameters for the normalized output.
+  MakeAlias(gammaTemp, W.memptr(), size, 1);
+  // Beta acts as the shifting parameters for the normalized output.
+  MakeAlias(betaTemp, W.memptr() + gammaTemp.n_elem, size, 1);
+
+  gammaTemp.fill(1.0);
+  betaTemp.fill(0.0);
+
+  runningMean.zeros(size, 1);
+  runningVariance.ones(size, 1);
+}
+
+template<typename MatType>
+void BatchNormType<MatType>::Forward(
+    const MatType& input,
+    MatType& output)
+{
+  const size_t batchSize = input.n_cols;
+  const size_t inputSize = inputDimension;
+
+  // We will calculate minibatch norm on each channel / feature map.
+  if (this->training)
+  {
+    // Check only during training, batch-size can be one during inference.
+    if ((batchSize * higherDimension) == 1 && inputSize == 1)
+    {
+      Log::Warn << "Variance for single element isn't defined and" <<
+          " will be set to 0.0 for training. Use a batch-size" <<
+          " greater than 1 to fix the warning." << std::endl;
+    }
+
+    // Input corresponds to output from previous layer.
+    // Used a cube for simplicity.
+    arma::Cube<typename MatType::elem_type> inputTemp(
+        const_cast<MatType&>(input).memptr(), inputSize, size,
+        batchSize * higherDimension, false, false);
+
+    // Initialize output to same size and values for convenience.
+    arma::Cube<typename MatType::elem_type> outputTemp(
+        const_cast<MatType&>(output).memptr(), inputSize, size,
+        batchSize * higherDimension, false, false);
+    outputTemp = inputTemp;
+
+    // Calculate mean and variance over all channels.
+    MatType mean = arma::mean(arma::mean(inputTemp, 2), 0);
+    variance = arma::mean(arma::mean(arma::pow(
+        inputTemp.each_slice() - arma::repmat(mean,
+        inputSize, 1), 2), 2), 0);
+
+    outputTemp.each_slice() -= arma::repmat(mean, inputSize, 1);
+
+    // Used in backward propagation.
+    inputMean.set_size(arma::size(inputTemp));
```

Comment:
This line happens implicitly in the `=` below, but also this is not the source of any bug. :)